protocols/mdns/src/behaviour/socket.rs

```
@@ -0,0 +1,190 @@
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Option<(usize, SocketAddr)>> {
+            match self.poll_recv_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_recv_from(buf) {
+                    Ok((len, from)) => {
+                        return Poll::Ready(Some((len, from)));
+                    }
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        return Poll::Ready(None);
+                    }
+                    Err(err) => {
+                        log::error!("Failed reading datagram: {}", err);
+                        return Poll::Ready(None);
+                    }
+                },
+                Poll::Ready(Err(e)) => {
+                    log::error!("Failed recv ready datagram: {}", e);
+                    return Poll::Ready(None);
+                }
+                _ => {}
+            }
+
+            Poll::Pending
+        }
+
+        /// Attempts to send data on the socket to a given address.
+        fn poll_send_packet(
+            &mut self,
+            cx: &mut Context,
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<()> {
+            match self.poll_send_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_send_to(packet, to) {
+                    Ok(_len) => {
+                        log::trace!("sent packet on iface {}", to);
+                        return Poll::Ready(());
+                    }
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        return Poll::Ready(());
+                    }
+                    Err(err) => {
+                        log::error!("Failed reading datagram: {}", err);
+                        return Poll::Ready(());
```

I am not sure about returning `Ready` here. Why not propagate the error?